skglm/solvers/base.py

```diff
@@ -1,8 +1,26 @@
-from abc import abstractmethod
+from abc import abstractmethod, ABC
+from skglm.utils.validation import check_obj_solver_attr
 
 
-class BaseSolver():
-    """Base class for solvers."""
+class BaseSolver(ABC):
+    """Base class for solvers.
+
+    Attributes
+    ----------
+    _datafit_required_attr : list
+        List of attributes that must implemented in Datafit.
+
+    _penalty_required_attr : list
+        List of attributes that must implemented in Penalty.
+
+    Notes
+    -----
+    For required attributes, if an attribute is given as a list of attributes
```

Comment:
I had trouble understanding this, an example help (in which case do we want to check that one of several attributes is present?)

Reply:
I 100% agree with you @mathurinm, I should have accompanied the docs with an example

In Fista solver, this

```python
    _datafit_required_attr = ("get_global_lipschitz", ("gradient", "gradient_scalar"))
    _penalty_required_attr = (("prox_1d", "prox_vec"),)
```

is interpreted as
- datafit is required to have ``get_global_lipschitz`` and (``gradient`` or ``gradient_scalar``)
- penalty is required to have  ``prox_1d`` or ``prox_vec``

This is the way I implemented ``check_obj_solver_attr`` function: whenever attributes are wrapped in parenthesis, it is interpreted as the “or” operator and a comma is interpreted the “and” operator.